src/java/org/apache/cassandra/cql3/statements/SelectStatement.java

```diff
@@ -288,7 +291,7 @@ public ResultMessage.Rows execute(QueryState state, QueryOptions options, long q
             query.trackWarnings();
         ResultMessage.Rows rows;
 
-        if (aggregationSpec == null && (pageSize <= 0 || (query.limits().count() <= pageSize)))
+        if (aggregationSpec == null && (pageSize <= 0 || (query.limits().count() <= pageSize) || query.isTopK()))
```

Comment:
`StorageAttachedIndex#validate` makes sure that we cannot use ANN without a `LIMIT`. However, paging is disabled in the CQL layer, so any index implementation could provide support for ANN. I think that the existence of a `LIMIT` should be checked here in `SelectStatement`.

As for the specific limit, 100 by default, I guess it could also be checked on the generic CQL layer because it affects the size of the pages. Should it also be checked by the specific index implementation?

Reply:
I don't think that we should be checking the actual maximum size of the limit here because that would be index specific but I have added a check that a limit has been specified and that it is `<=` to the page size (if the page size is defined).